Bluetooth: controller: Fix handling of ext adv HCI commands

andrzej-kaczmarek opened the pull request:
These patches fix handling of ext adv HCI commands, specifically mixing legacy/extended commands and handling of advertising handle parameter.

andrzej-kaczmarek committed:
Bluetooth: controller: Disallow mixing legacy/ext adv HCI commands

As per Core 5.2, Part B, section 3.1.1 controller should not allow
mixing of legacy and extended advertising commands. Once 1st command
is received from host, we should only allow commands of the same type
and return an error on other commands.

If legacy advertising interface is selected by 1st command received,
we need to make sure that set with handle=0 is created since it's
used by all legacy commands.

When not using external host, we assume that only extended advertising
interface will be used thus all checks can be skipped (optimized and
compiled-out).

Signed-off-by: Andrzej Kaczmarek <andrzej.kaczmarek@codecoup.pl>

andrzej-kaczmarek committed:
Bluetooth: controller: Fix set_ext_adv_param returned status

Should not always return success.

Signed-off-by: Andrzej Kaczmarek <andrzej.kaczmarek@codecoup.pl>

andrzej-kaczmarek committed:
Bluetooth: controller: Remove redundant flags for advertising set

There's not need to mark advertising set as created by either legacy
or advertising command since this is determined by current so-called
HCI advertising mode which determines whether host uses legacy or
extended advertising commands.

Signed-off-by: Andrzej Kaczmarek <andrzej.kaczmarek@codecoup.pl>

andrzej-kaczmarek committed:
Bluetooth: controller: Fix return value type

This should match prototype in ll.h.

Signed-off-by: Andrzej Kaczmarek <andrzej.kaczmarek@codecoup.pl>

andrzej-kaczmarek committed:
Bluetooth: controller: Add advertising set handle mapping

Core 5.2, Vol 4, Part E, section 5.3.1 clarifies that advertising set
handle is assigned by host when advertising set is created and this
happens only on LE Set Extended Advertising Parameters.

An advertising set handle is an arbitrary number within allowed range,
i.e. 0x00-0xEF and not 0..max_supported-1.

This patch adds option to enable advertising set handle mapping from
HCI range as defined by Core specification to zero-based handles used
by LL. If enabled, HCI handle will be remapped to LL handle for each
command, otherwise HCI handle will be used as an LL handle. The latter
effectively skips mapping logic and should be used with Zephyr host
which uses zero based indexes.

Signed-off-by: Andrzej Kaczmarek <andrzej.kaczmarek@codecoup.pl>

cvinayak commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -955,6 +990,10 @@ static void le_set_adv_param(struct net_buf *buf, struct net_buf **evt)
 	uint16_t min_interval;
 	uint8_t status;
 
+	if (hci_adv_mode_ensure_legacy(evt)) {
```
```suggestion
	if (IS_ENABLED(CONFIG_BT_HCI_RAW) && hci_adv_mode_ensure_legacy(evt)) {
```

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_adv.c:
```diff
@@ -81,6 +81,23 @@ static struct ll_adv_set ll_adv[BT_CTLR_ADV_SET];
 static struct ticker_ext ll_adv_ticker_ext[BT_CTLR_ADV_SET];
 #endif /* CONFIG_BT_TICKER_EXT */
 
+#if defined(CONFIG_BT_CTLR_ADV_EXT)
```
```suggestion
#if defined(CONFIG_BT_HCI_RAW) && defined(CONFIG_BT_CTLR_ADV_EXT)
```

cvinayak commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -1662,6 +1729,10 @@ static void le_set_adv_set_random_addr(struct net_buf *buf,
 	struct bt_hci_evt_cc_status *ccst;
 	uint8_t status;
 
+	if (hci_adv_mode_ensure_ext(evt)) {
```
```suggestion
	if (IS_ENABLED(CONFIG_BT_HCI_RAW) && hci_adv_mode_ensure_ext(evt)) {
```

cvinayak commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -107,6 +107,41 @@ static uint64_t event_mask = DEFAULT_EVENT_MASK;
+	return err;
+}
+#else
+static int hci_adv_mode_ensure_legacy(struct net_buf **evt)
```
```suggestion
static inline int hci_adv_mode_ensure_legacy(struct net_buf **evt)
```

cvinayak commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -107,6 +107,41 @@ static uint64_t event_mask = DEFAULT_EVENT_MASK;
 static uint64_t event_mask_page_2 = DEFAULT_EVENT_MASK_PAGE_2;
 static uint64_t le_event_mask = DEFAULT_LE_EVENT_MASK;
 
+#if defined(CONFIG_BT_CTLR_ADV_EXT)
```
```suggestion
#if defined(CONFIG_BT_HCI_RAW) && defined(CONFIG_BT_CTLR_ADV_EXT)
```

cvinayak replied:
stale

cvinayak reviewed (changes requested):
Avoid HCI checks when Zephyr BT host used. HCI parameter and protocol checks enabled only when standalone controller is built.

cvinayak commented on subsys/bluetooth/controller/include/ll.h:
```diff
@@ -7,13 +7,18 @@
 
 #define LL_VERSION_NUMBER BT_HCI_VERSION_5_2
 
+#define LL_ADV_HCI_MODE_ANY    0 /* Any advertising cmd/evt allowed */
+#define LL_ADV_HCI_MODE_LEGACY 1 /* Only legacy advertising cmd/evt allowed */
+#define LL_ADV_HCI_MODE_EXT    2 /* Only extended advertising cmd/evt allowed */
+
```
```suggestion
#define LL_ADV_INTF_ANY    0 /* Any advertising cmd/evt allowed */
#define LL_ADV_INTF_LEGACY 1 /* Only legacy advertising cmd/evt allowed */
#define LL_ADV_INTF_EXT    2 /* Only extended advertising cmd/evt allowed */

```

cvinayak commented on subsys/bluetooth/controller/include/ll.h:
```diff
@@ -7,13 +7,18 @@
 uint8_t ll_addr_set(uint8_t addr_type, uint8_t const *const p_bdaddr);
 
 #if defined(CONFIG_BT_CTLR_ADV_EXT)
+int ll_adv_hci_mode_set(uint8_t mode);
```
```suggestion
int ll_adv_intf_set(uint8_t mode);
```

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_adv.c:
```diff
@@ -81,6 +81,23 @@ static struct ll_adv_set ll_adv[BT_CTLR_ADV_SET];
 #endif /* CONFIG_BT_TICKER_EXT */
 
+#if defined(CONFIG_BT_CTLR_ADV_EXT)
+static uint8_t ll_adv_hci_mode;
```
```suggestion
static uint8_t ll_adv_intf;
```

cvinayak commented on subsys/bluetooth/controller/include/ll.h:
```diff
@@ -19,6 +19,7 @@ uint8_t ll_addr_set(uint8_t addr_type, uint8_t const *const p_bdaddr);
 
 #if defined(CONFIG_BT_CTLR_ADV_EXT)
 int ll_adv_hci_mode_set(uint8_t mode);
+int ll_adv_hci_mode_is_ext(void);
```
```suggestion
int ll_adv_intf_is_ext(void);
```

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_adv.c:
```diff
@@ -1213,16 +1236,50 @@ int ull_adv_reset(void)
 inline uint16_t ull_adv_handle_get(struct ll_adv_set *adv)
 {
-	return ((uint8_t *)adv - (uint8_t *)ll_adv) / sizeof(*adv);
+#if defined(CONFIG_BT_CTLR_ADV_SET)
+	return adv->handle;
+#else
+	LL_ASSERT(adv == &ll_adv[0]);
+	return 0;
+#endif
```
Zero-indexed handles is used as ticker instance offset, I think. Move HCI handle lookup as separate feature in HCI module.

jhedberg commented on subsys/bluetooth/controller/ll_sw/ull_adv_types.h:
```diff
@@ -19,6 +19,7 @@ struct ll_adv_set {
 	uint8_t  rnd_addr[BDADDR_SIZE];
 	uint8_t  sid:4;
 	uint8_t  is_created:1;
+	uint8_t  handle;
```
> Thinking aloud, maybe we also need a new Kconfig in combination to CONFIG_BT_HCI_RAW, a CONFIG_BT_HCI_ZEPHYR for multi-core zephyr eco system to indicate the host on the other core is Zephyr.

I'd call it HCI_TRUSTED or similar, since essentially you're saying that you trust the host to only feed you valid data.

carlescufi commented:
@andrzej-kaczmarek will you follow-up on the reivew comments?

andrzej-kaczmarek committed:
Bluetooth: controller: Fix error code on invalid adv set handle

Proper error code should be returned if advertising set with given
handle does not exist instead of generic "invalid parameters" one.

Signed-off-by: Andrzej Kaczmarek <andrzej.kaczmarek@codecoup.pl>

andrzej-kaczmarek committed:
Bluetooth: controller: Add helper for CC event with status only

A lot of commands return CC event with status parameter only, especially
when error is returned. This patch adds a helper to create such event,
similar to cmd_status.

Signed-off-by: Andrzej Kaczmarek <andrzej.kaczmarek@codecoup.pl>

andrzej-kaczmarek commented:
@cvinayak @carlescufi 
sorry it took so long to update. all comments should be now fixed. I made some more changes in this PR to address issues I found on my side and avoid too much c&p code.

I decided to skip #ifdefs for most code and use static inlines instead which are optimized and compiled-out if certain options are disabled, it makes much less c&p code.

there was also problem that invalid HCI event was used (CC vs. CS) for some HCI commands, this is now also fixed.

cvinayak commented on subsys/bluetooth/controller/Kconfig:
```diff
@@ -94,6 +94,18 @@ config BT_CTLR_HCI_VS_BUILD_INFO
+	bool "Enable advertising set handle mapping between HCI and LL"
+	default y if BT_HCI_RAW
+	default n
+	depends on BT_CTLR_ADV_EXT
```
nitpick, move this line below the `bool` prompt like, so that `depends on` is close to the `config` even when additions defaults are appended, if any, in the future.

cvinayak reviewed (changes requested):
Except for couple minor nitpick, changes look good to me.

carlescufi commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -103,6 +103,55 @@ static uint64_t event_mask = DEFAULT_EVENT_MASK;
+
+#if defined(CONFIG_BT_CTLR_ADV_EXT)
+#if defined(CONFIG_BT_HCI_RAW)
+static int adv_intf_ensure_legacy(struct net_buf **cc_evt)
```
Can we rename those functions to something a bit clearer
Perhaps

```
ll_adv_cmds_set()
adv_cmds_ensure_<>()
```

andrzej-kaczmarek replied:
> Can we rename those functions to something a bit clearer
> Perhaps
> 
> ```
> ll_adv_cmds_set()
> adv_cmds_check(bool legacy, ...)
> ```

I'm fine with either version and don't have any better name for that setting, so if @cvinayak also agrees I'll just change in next version

cvinayak replied:
@carlescufi can we agree to have two explicit functions? `adv_cmd_legacy_check(struct net_buf **cc_evt)` and `adv_cmd_ext_check(struct net_buf **cc_evt)`; this will avoid the additional parameter passing, unless all callers pass a constant value and are all in the same file. (my belief is link time optimizations is not present in zephyr?).

carlescufi replied:
@cvinayak this will increase code size, LTO is not enabled in Zephyr. I'd rather have a single function to be honest.

cvinayak replied:
@carlescufi no strong objections, I see @andrzej-kaczmarek has split into two separate functions, if he decides that there is not code size increase, I will go with his choice/decision.

andrzej-kaczmarek commented:
not quite sure why buildkite failed, does not seem to be due to any build error...

andrzej-kaczmarek commented on subsys/bluetooth/controller/ll_sw/ull_adv.c:
```diff
@@ -192,6 +241,9 @@ uint8_t ll_adv_params_set(uint16_t interval, uint8_t adv_type,
 
 	adv->is_created = 1;
+#if defined(CONFIG_BT_CTLR_HCI_ADV_HANDLE_MAPPING)
+	adv->hci_handle = handle;
```
this won't work since currently `handle` passed here is still a ULL handle, not HCI handle.
I think easiest way to fix this would be to populate `hci_handle` in `ll_adv_set_by_hci_handle_get_or_new` since it will not be used anyway until `is_created` is set so will update.

cvinayak reviewed (approved):
No further requests from me. QUICK-ACK on the latest code in this PR. Approved after a quick look.

carlescufi commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -129,6 +129,18 @@ static struct net_buf *cmd_status(uint8_t status)
+	struct net_buf *buf;
+	struct bt_hci_evt_cc_status *ccst;
+
+	buf = bt_hci_cmd_complete_create(_opcode, sizeof(*ccst));
```
@andrzej-kaczmarek why are you not using `hci_cmd_complete()` here instead of duplicating the code?

carlescufi replied:
Could you check both options and see if there´s an impact to code size?

andrzej-kaczmarek replied:
tbh I just did not think about it, but have just checked both versions and current one is a bit smaller:
`FLASH:      127124 B         1 MB     12.12%` (current)
`FLASH:      127588 B         1 MB     12.17%` (fixed)
numbers above are from `hci_usb`

carlescufi commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -103,6 +103,55 @@ static uint64_t event_mask = DEFAULT_EVENT_MASK;
+
+#if defined(CONFIG_BT_CTLR_ADV_EXT)
+#if defined(CONFIG_BT_HCI_RAW)
+static int adv_cmds_legacy_check(struct net_buf **cc_evt)
```
Wouldn't it be nicer to have a single set of functions and then have the `#if defined(CONFIG_BT_HCI_RAW)` checks inside the functions?

andrzej-kaczmarek replied:
indeed, will fix

carlescufi reviewed (commented):
Very nice patchset thank you. Just a couple of comments
